rs/cli/src/auth.rs

```diff
@@ -156,10 +106,60 @@ impl Auth {
         }
         Ok(None)
     }
+}
 
-    pub async fn auto_detect_neuron_id(&self, nns_urls: &[url::Url]) -> anyhow::Result<u64> {
-        let sender = match self {
-            Auth::Hsm { pin, slot, key_id } => {
+#[derive(Clone, Debug)]
+pub struct Auth {
+    pub auth_type: AuthType,
+    maybe_neuron: Option<Neuron>,
+}
+
+impl Auth {
+    pub fn new(auth_type: AuthType) -> Self {
+        Self {
+            auth_type,
+            maybe_neuron: None,
+        }
+    }
+
+    pub fn with_neuron(self, neuron: Neuron) -> Self {
+        Self {
+            maybe_neuron: Some(neuron),
+            ..self
+        }
+    }
+
+    pub async fn with_autodetect_neuron(self, network: &Network) -> anyhow::Result<Self> {
+        let neuron = Neuron {
+            neuron_id: self.auto_detect_neuron_id(&network.nns_urls).await?,
+            include_proposer: true,
+        };
+
+        Ok(Self {
+            maybe_neuron: Some(neuron),
+            ..self
+        })
+    }
```

Comment:
Similar as above. Better to provide one constructor and disallow the calls that will take ownership and transform. If there is a quick fix people will use it